citation/ping_urls.py

```diff
@@ -0,0 +1,94 @@
+import logging
+import requests
+
+from django.db.models import Q
+
+from citation.models import Publication, URLStatusLogs, PlatformTypes
+
+logger = logging.getLogger(__name__)
+
+""" 
+    Contains definition that are used to 
+       - test the code archive urls are valid and up 
+       - Categorize them 
+       - Logs them in the URLStatusLog table
+
+"""
+
+
+def verify_url_status():
+    publication = Publication.objects.filter(~Q(code_archive_url = ''), is_primary = "True", status= "REVIEWED")
+
+    for pub in publication:
+        logger.debug("Pinging: " + pub.code_archive_url)
+        ping_url(pub.pk, pub.code_archive_url)
+
+def ping_url(pk, url):
+    try:
+        s = requests.Session()
+        r = requests.Request('GET', url)
+        resp = s.send(r.prepare())
+        resp.raise_for_status()
+        platform_type = categorize_url(url)
+        add_url_status_log(pk, platform_type, resp)
+
+    except requests.exceptions.HTTPError as err:
+        # URL Not found or forbidden (Private access)
+        add_url_status_log(pk, PlatformTypes.INVALID, err.response)
+
+    except requests.exceptions.RequestException as e:
+        # Server not reachable
+        add_url_status_log_bad_request(pk, PlatformTypes.INVALID, url)
+
+"""
+    Categorize the url depending on the server name into following categories 
+    CoMSES, Open Source, Platforms, Journal, Personal, Others, and Invalid
+"""
+# FIXME This may not be the best way to categorize url - in future categorization will be added manually while updating
+def categorize_url(url):
```

Comment:
This nested chain of if elifs should be refactored. One suggestion: create a `UrlParser` class or `url_parse` method that takes in a URL and returns the appropriate `PlatformType`.

At a minimum you should extract the logic out into methods like `if is_comses_url(url): ... elif is_oss_url(url): ...`. Let me know if you have any questions about this, happy to clarify further.